test/karma/spec/loginFlow.js

```diff
@@ -277,4 +279,56 @@ describe('Complete login flow', function() {
     });
   });
 
+  it('PKCE: throws when authorization code is expired', function() {
+    // First hit /authorize
+    return bootstrap()
+    .then(function(app) {
+      mockWellKnown();
+      return app.loginRedirect({
+        pkce: true,
+        nonce: NONCE,
+        codeVerifier: CODE_VERIFIER
+      });
+    }).then(function() {
+      return waitFor(function() {
+        return setLocation.calls.any();
+      });
+    })
+    .then(function() {
+      var url = new URL(setLocation.calls.first().args[0]);
+      return url;
+    })
+    .then(function(url) {
+      jasmine.Ajax.requests.reset();
+
+      // Now we handle the redirect & hit /token
+      const state = url.searchParams.get('state');
+      const pathname = `${CALLBACK_PATH}?code=${AUTHORIZATION_CODE}&state=${state}`;
+
+      // simulate authorization code expiration
+      setTimeout(function () {
+        jasmine.Ajax.stubRequest(
+          /.*v1\/token/
+        ).andReturn({
+          status: 400,
+          responseJSON: {
+            'error': 'invalid_grant',
+            'error_description': 'The authorization code is invalid or has expired.'
+          }
+        });
+      }, 60000);
+
+      return Promise.resolve().then(function () {
+        jasmine.clock().tick(60000);
```

Comment:
is this necessary? We are simulating a server response, seems like it should not depend on the client clock?

Reply:
I believe it is not. The intent was to mimic the delay between `/authorize` and `/token` calls although there is no extra value in doing so.